[Messenger][Process] add `RunProcessMessage` and `RunProcessMessageHandler`

kbond opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | n/a
| License       | MIT
| Doc PR        | todo

When reviewing how I could replace my cron scheduler with symfony/scheduler, I realized I need a way to schedule processes.

## Usage

```php
use Symfony\Component\Process\Exception\RunProcessFailedException;
use Symfony\Component\Process\Messenger\RunProcessMessage;

try {
    $context = $bus->dispatch(new RunProcessMessage('something'));

    $context->exitCode; // int
    $context->output; // string
    $context->errorOutput; // string
catch(RunProcessFailedException $e) {
    $e->context->exitCode; // int
    $e->context->output; // string
    $e->context->errorOutput; // string
}
```

TODO:
- [x] wire up
- [x] tests

kbond committed:
[Messenger][Process] add `RunProcessMessage` and `RunProcessMessageHandler`

kbond commented on src/Symfony/Component/Process/Messenger/ExecuteProcessHandler.php:
```diff
@@ -0,0 +1,33 @@
+            $message->timeout,
+        );
+
+        return $process->mustRun();
```
Throws an exception if the process fails, should this be configurable?

kbond replied:
This is actually going to be a problem I think. `ProcessFailedException` (which this line throws on error) can't be serialized because it contains the `Process`.

Like stated in https://github.com/symfony/symfony/pull/49814#issuecomment-1486903588, it is useful to have the output in the exception. Should we create a specific exception for this (`RunProcessFailedException`) that has the exit code, output and error output?

kbond commented:
Looking for feedback before proceeding further.

fabpot commented on src/Symfony/Component/Process/Messenger/ExecuteProcess.php:
```diff
@@ -0,0 +1,27 @@
+/**
+ * @author Kevin Bond <kevinbond@gmail.com>
+ */
+final class ExecuteProcess
```
What about `ProcessMessage` and `ProceMessageHander`?
It looks like you're going to fix all my wish list pretty fast ;)
https://speakerdeck.com/fabpot/s

kbond replied:
Hah, I did read that in your slide but didn't make the connection that this is what you were referring to!

OskarStark replied:
ExecuteProcessMessage ?

kbond replied:
At a second glance `ProcessMessage` does read like _processing a message_...

OskarStark replied:
That's why I proposed ExecuteProcessMessage, "ExecuteProcess" (what it did) and "Message" is just the suffix

nicolas-grekas replied:
ExecuteProcessMessage and ExecuteCommandMessage, or RunProcessMessage/RunCommandMessage?

OskarStark replied:
I like the "Run" prefix 👍

OskarStark commented on src/Symfony/Component/Process/CHANGELOG.md:
```diff
@@ -1,6 +1,11 @@
+6.3.0
+-----
+
+ * added `ExecuteProcess` and `ExecuteProcessHandler`
```
```suggestion
 * Add `ExecuteProcess` and `ExecuteProcessHandler`
```

OskarStark commented on src/Symfony/Component/Process/CHANGELOG.md:
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+6.3.0
```
```suggestion
6.3
```

OskarStark commented on src/Symfony/Component/Process/CHANGELOG.md:
```diff
@@ -1,6 +1,11 @@
 =========
 
+6.3.0
+-----
```
```suggestion
---
```

fabpot commented on src/Symfony/Component/Process/Messenger/ProcessMessage.php:
```diff
@@ -0,0 +1,32 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Process\Messenger;
```
I would move it under the Messenger namespace instead. I think it will make discoverability a bit easier.

kbond replied:
Ok, should I do the same for #49815 and #49814?

ro0NL commented on src/Symfony/Component/Process/Messenger/ProcessMessage.php:
```diff
@@ -0,0 +1,32 @@
+    ) {
+    }
+
+    public function __toString(): string
```
generally what about just dumping payloads (`$dump($messsage)`) in CLI for -v[vv]?

kbond replied:
Can you explain this a bit more, I'm not understanding what you mean.

ro0NL replied:
https://github.com/symfony/console/blob/aa5d64ad3f63f2e48964fc81ee45cb318a723898/Helper/Dumper.php#L22

`$dump = new Dumper($output);`

kbond commented:
I've renamed the classes per the review.

I've also added a `RunProcessContext` that's returned from the handler and a `RunProcessFailedException` that's thrown by the handler on failure. These will allow the message/exception to be serialized while still giving access to the result.

kbond commented on src/Symfony/Component/Process/Messenger/RunProcessFailedException.php:
```diff
@@ -0,0 +1,26 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Process\Messenger;
```
Should this be under the `Process\Exception` namespace?

stloyd replied:
I think this is quite common in the codebase, so I think yes

stloyd commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -240,6 +241,12 @@ public function load(array $configs, ContainerBuilder $container)
 
+        $loader->load('process.php');
+
+        if (!class_exists(RunProcessMessageHandler::class)) {
```
Maybe put load into the if statement and remove negation? Why load if we know there is no such class?

kbond replied:
I followed the standard of the rest of this file.

stloyd commented on src/Symfony/Component/Process/Messenger/RunProcessMessage.php:
```diff
@@ -0,0 +1,32 @@
+        public readonly ?string $cwd = null,
+        public readonly ?array $env = null,
+        public readonly mixed $input = null,
+        public readonly ?float $timeout = 60,
```
60.0? As we expect float?

kbond commented:
Comments addressed and PR description updated to show usage.

nicolas-grekas commented on src/Symfony/Component/Process/Messenger/RunProcessContext.php:
```diff
@@ -0,0 +1,33 @@
+/**
+ * @author Kevin Bond <kevinbond@gmail.com>
+ */
+final class RunProcessContext extends RunProcessMessage
```
```suggestion
class RunProcessContext extends RunProcessMessage
```

nicolas-grekas commented on src/Symfony/Component/Process/Messenger/RunProcessMessageHandler.php:
```diff
@@ -0,0 +1,39 @@
+{
+    public function __invoke(RunProcessMessage $message): RunProcessContext
+    {
+        $process = Process::fromShellCommandline(
```
I missed this in previous reviews: we shouldn't use fromShellCommandline, it's a pending door to command line injection.
Instead we should define commands as arrays.
We might also support shell command lines, but these should be 2nd class.

kbond replied:
Good call.

fabpot commented:
Thank you @kbond.
